suite/tests/client-interface/cleancallregs.dll.c

```diff
@@ -0,0 +1,270 @@
+/* **********************************************************
+ * Copyright (c) 2017 ARM Limited. All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of ARM Limited nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL ARM LIMITED OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* Marshalling the arguments to a clean call is non-trivial when the arguments
+ * are passed in registers and the values in some registers need to be permuted
+ * (and we wish to do this efficiently). This test checks some interesting cases.
+ *
+ * It makes no difference what app is used. The first basic block encountered will
+ * be instrumented with a sequence of clean calls followed by a call via finish()
+ * to dr_exit_process(), so no app code is executed.
+ *
+ * This test can easily be ported to a new architecture. There is also the
+ * X86-specific test "cleancall".
```

Comment:
It doesn't seem hard to merge to me: the arg combos here are self-contained, right?  It seems like they can easily be plunked down as yet another isolated case in the existing test.  The existing test expects a crash, so the current exit can just be changed to that.

OTOH that requires building the current test for ARM, probably by putting ifdefs around some of the x86-specific pieces inside it or making them cross-platform, which is extra work (though work we'd like to do eventually).

From a test suite efficiency standpoint, having fewer but larger executable takes less time to build and run: not sure that should be a big concern though.

Up to you.  Just tweaking the comment is the simplest path forward.  Re: names: the existing test tests spilling regs to tls and access them from the clean call, along with reg params, so it does handle "clean call regs", but I don't know what else to call this test: maybe "cleancallparams" is closer as it is only concerned with direct params.